Introduced an STL container for a view (util::view<T>), and changed opresolv to use it

npwoods opened the pull request:
Since this involves creating an STL container from scratch, this is another case where I want the hardcore C++ people to review things with a fine toothed comb.

cuavas commented on src/lib/util/view.h:
```diff
@@ -0,0 +1,154 @@
+
+#pragma once
+
+#ifndef __VIEW_H__
```
I'm sick of saying it: this is unacceptable.  **ALL NAMES BEGINNING WITH AN UNDERSCORE OR CONTAINING TWO CONSECUTIVE UNDERSCORES ARE RESERVED IN C++.**
Use something like `MAME_LIB_UTIL_VIEW_H`.

cuavas commented:
“View” is a terrible name for this.  Call it `contiguous_sequence_wrapper` or something.

cuavas commented on src/lib/util/view.h:
```diff
@@ -0,0 +1,154 @@
+		typedef std::random_access_iterator_tag iterator_category;
+
+		iterator(T *ptr) : m_ptr(ptr) { }
+		iterator(const iterator &that) : m_ptr(that) { }
```
You can use `= default` or not declare this since it just does a memberwise copy.  Same for the assignment operator.

cuavas commented on src/lib/util/view.h:
```diff
@@ -0,0 +1,154 @@
+		difference_type operator-(const iterator &that) const { return m_ptr - that.m_ptr; }
+
+		iterator& operator++() { m_ptr++; return *this; }
+		iterator operator++(int) { iterator result = *this; m_ptr++; return result; }
```
You should add matching pre-/post-decrement operators to meet the requirements of [BidirectionalIterator](http://en.cppreference.com/w/cpp/concept/BidirectionalIterator) which is a base concept of [RandomAccessIterator](http://en.cppreference.com/w/cpp/concept/RandomAccessIterator).

cuavas commented on src/lib/util/view.h:
```diff
@@ -0,0 +1,154 @@
+
+		reference operator*() const { return *m_ptr; }
+		pointer operator->() const { return m_ptr; }
+		operator T*() const { return m_ptr; }
```
Implicit cast operators are evil, just don't do it.  If you wanted to use naked pointers, you'd just declare the iterator to be a pointer (`std::iterator_traits` works on pointers).

ajrhacker replied:
The implicit cast operator also appears to be the cause of the "ambiguous conversion" error in the tea-ci build.

cuavas commented on src/lib/util/view.h:
```diff
@@ -0,0 +1,154 @@
+	{
+	}
+
+	view(const view &that)
```
As with the iterators, there's no point doing this explicitly, just use `= default` to get the standard copy constructor.

cuavas commented on src/lib/util/view.h:
```diff
@@ -0,0 +1,154 @@
+
+		typedef std::random_access_iterator_tag iterator_category;
+
+		iterator(T *ptr) : m_ptr(ptr) { }
```
Could you qualify this constructor `explicit`?  It makes it more obvious when people are trying to convert one to the other.

cuavas commented on src/lib/util/view.h:
```diff
@@ -0,0 +1,154 @@
+	bool empty() const { return size() > 0; }
+
+	// element access
+	reference front() { return *this[0]; }
```
This is very, very wrong.  What's the precedence of suffix vs prefix operators in C++?  It's effectively `*(this[0])` when you want `(*this)[0]` (you're subscripting the `this` pointer, not the object).  It's clearer to write `return operator[](0)` anyway.

cuavas commented:
You're missing the `iterator operator+(iterator::difference_type, iterator)` operators for `iterator` and `const_iterator` needed for [RandomAccessIterator](http://en.cppreference.com/w/cpp/concept/RandomAccessIterator) - these need to be declared as free functions outside the class, not members.

cuavas commented:
Have you considered implementing `rbegin`/`rend` if nothing else for the exercise?

npwoods commented:
I think I've addressed all of these concerns, with the exception of implementing rbegin() and rend().  What is the best way to do so?  I strongly suspect that creating another reverse_iterator and const_reverse_iterator objects are not the way to go.

ajrhacker commented:
One remaining bug is the use of m_ptr (which no longer exists) in operator[]. RandomAccessIterator traits are already assumed by the code, so `return m_begin[n]` will suffice here.

(The filename also seems a bit long now, but that may just be me. Using the full name of the class seems like a Javaism.)

npwoods commented:
Good catch - I've removed the m_ptr checks

The thought that the filename was too long occurred to me; I couldn't think of a "good" way to make it smaller.  If anyone has any ideas, just let me know.

cuavas commented on src/lib/util/contiguous_sequence_wrapper.h:
```diff
@@ -0,0 +1,89 @@
+	std::reverse_iterator<const_iterator> crbegin() const { return std::reverse_iterator<const_iterator>(cend() - 1); }
+	std::reverse_iterator<iterator> rend() { return std::reverse_iterator<iterator>(begin() - 1); }
+	std::reverse_iterator<const_iterator> rend() const { return std::reverse_iterator<iterator>(begin() - 1); }
+	std::reverse_iterator<const_iterator> crend() const { return std::reverse_iterator<iterator>(begin() - 1); }
```
Have you actually tested this?  I'm pretty sure you're not supposed to use all the -1 because of how the reverse iterator actually works: it wraps a forward iterator one beyond the element it points to.

cuavas commented on src/lib/util/contiguous_sequence_wrapper.h:
```diff
@@ -0,0 +1,89 @@
+	// capacity
+	size_type size() const { return m_end - m_begin; }
+	size_type max_size() const { return size(); }
+	bool empty() const { return size() > 0; }
```
Your logic is backwards.

cuavas commented:
The template isn't huge, maybe it could go in src/lib/util/coretmpl.h so that file becomes less useless?

cuavas commented on src/lib/util/coretmpl.h:
```diff
@@ -351,4 +351,74 @@ class fixed_allocator
+namespace util {
+
+// wraps an existing sequence of values
+template<typename T >
```
The spacing is screwed up - how about using `template <typename T>` to be more like other code?

cuavas commented on src/lib/util/coretmpl.h:
```diff
@@ -351,4 +351,74 @@ class fixed_allocator
+	typedef const T &const_reference;
+	typedef T *pointer;
+	typedef T *iterator;
+	typedef const T *const_iterator;
```
Maybe add a typedef for `reverse_iterator` and its const friend?

cuavas commented on src/lib/util/coretmpl.h:
```diff
@@ -351,4 +351,74 @@ class fixed_allocator
+	// capacity
+	size_type size() const { return m_end - m_begin; }
+	size_type max_size() const { return size(); }
+	bool empty() const { return size() > 0; }
```
This is still wrong.

cuavas commented on src/lib/util/coretmpl.h:
```diff
@@ -351,4 +351,74 @@ class fixed_allocator
+	void check_in_bounds(size_type n)
+	{
+		if (n < 0 || n >= size())
+			throw new std::out_of_range("invalid contiguous_sequence_wrapper<T> subscript");
```
What are you?  A C# programmer?
